### PR TITLE
Fix missing

### LINKED ISSUE #14: Missing student Name in the output file for specific cases
We found a case where the rubric data contained 2 `artifact_ids` that do not exist (Ex: submission_dim). These artifact_ids only exist in `rubrics.assessments`.

It's possible that StudentView "submission" created an artifact.

### LINKED ISSUE #15: Missing students who do have rubric data
We found a case where there are some students who do have rubric data who are missing in the script's final output.

They have inputted scores of 0 which differ from the rubric.

### LINKED ISSUE #16: Missing students who did not take the midterm / assignment
We have a case where there seem to be some students who do not have rubric scores. This is different from a non-submission.

### LINKED ISSUE #17: Script specific to TA as grader
The script assumes the grader will be a TA role - but there are other options (including instructor, secondary instructor etc.). Update to user "grader" who is anyone that is not a student.

### PR DESCRIPTION
**Pull request by alisonmyers**
User identified
1) students missing from output (no rubric score given)  closes #16 
2) rubric data not associated with any individual student closes #14 
3) missing students with rubric data closes #15 

Also noticed
4) scripts are specific to TAs as graders closes #17 

Changed order of operations / join conditions
1) The original script uses the rubric.assessments as the initial join condition - update uses student enrollments
2) Additional fields added for clarity: submission state, and submission score
3) Grader not specific to TAs only.

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -57,7 +57,7 @@ const getAssignmentRubricAndWriteFile = async (courseId, assignmentId, rubricId)
   console.log('Thanks, fetching data now. This may take a bit of time...')
   const rubricData = await getRubric(courseId, assignmentId, rubricId)
   writeToCSV(rubricData, `rubric-${courseId}-${assignmentId}-${rubricId}.csv`)
-  console.log(`Data is ready now. It's named 'rubric-${courseId}-${assignmentId}-${rubricId}.csv'`)
+  //console.log(`Data is ready now. It's named 'rubric-${courseId}-${assignmentId}-${rubricId}.csv'`)
```
Comment (justin0022) [inline cli.js]: did you intentionally comment this out?

File: getRubric.js [hunk 2]
```diff
@@ -2,64 +2,74 @@ const { flatten } = require('ramda')
...
-      .map(({ points, comments }) => ({ points, comments }))
+    const userName = user.name || ''
+    const userSISID = user.sis_user_id || ''
+    const userCanvasID = user.id || ''
 
-    // link to assignment
-    const url = submission.attachments
-      ? submission.attachments[0].url
-      : ''
+    const section = sections.find(student => student.course_section_id == sections.id ) || {}
```
Comment (justin0022) [inline getRubric.js]: should be `===`

File: getRubric.js [hunk 3]
```diff
@@ -2,64 +2,74 @@ const { flatten } = require('ramda')
...
-    const submission = submissions
-      .find(submission => submission.id === submissionId) || {}
-    const studentId = submission.user_id || ''
-    const student = sections
-      .find(student => student.id === studentId) || {}
-    const studentName = student.name || ''
-    const studentNumber = student.sis_user_id || ''
-    const section = student.section || ''
+  const students = enrollments.filter(enrollment => enrollment.type == 'StudentEnrollment') // excludes StudentViewEnrollment but keeps StudentEnrollment, to keep studentView use enrollment.role
+  const nonStudents = enrollments.filter(enrollment => enrollment.role != 'StudentEnrollment') // excludes both StudentViewEnrollment and StudentEnrollment
```
Comment (justin0022) [inline getRubric.js]: should be `!==`

File: getRubric.js (hunk 2 above)
Comment (justin0022) [reply getRubric.js]: all `==` should be updated to `===`

File: getRubric.js [hunk 4]
```diff
@@ -2,64 +2,74 @@ const { flatten } = require('ramda')
...
+    canvasAPI.getAssignmentSubmissions(courseId, assignmentId, canvasAPI.getOptions.submissions.submission_comments, canvasAPI.getOptions.submissions.rubric_assessment),
     canvasAPI.getRubric(courseId, rubricId),
-    canvasAPI.getSections(courseId, canvasAPI.getOptions.users.include.students)
-      .then(sections => flatten(
-        sections
-          .map(({ name, students }) => students
-            .map(student => ({ ...student, section: name }))
-          )
-      ))
+    canvasAPI.getSections(courseId)
```
Comment (justin0022) [inline getRubric.js:8]: I noticed that you got rid of the:

```js
 .then(sections => flatten(
        sections
          .map(({ name, students }) => students
            .map(student => ({ ...student, section: name }))
          )
      ))
```
Just wanted to double check that this is intentional?

File: getRubric.js [hunk 5]
```diff
@@ -2,64 +2,74 @@ const { flatten } = require('ramda')
...
+    ? submission.submission_comments
+      .filter(comment => comment.author.id !== userCanvasID)
+      .map(comment => comment.comment)
+    : []
+  
+    const rubricAssessments = rubrics.assessments.find(rubric => rubric.artifact_id == submissionId) || {}
+
+    const rubricScore = rubricAssessments.score || ''
+
+    const rubricData = (rubricAssessments.data || []).map(({ points, comments }) => ({ points, comments }));
```
Comment (justin0022) [inline getRubric.js]: remove semicolon

File: cli.js (hunk 1 above)
Comment (alisonmyers) [reply cli.js]: no ... will fix.

File: getRubric.js (hunk 4 above)
Comment (alisonmyers) [reply getRubric.js:8]: Yes - we just need the section name

**Comment by alisonmyers** [comment]
Updated based on review feedback.

**Comment by alisonmyers** [comment]
Found new issue - need to account for rubric columns for a completely empty rubric before merging

File: getRubric.js [hunk 6]
```diff
@@ -1,65 +1,73 @@
...
-      .find(submission => submission.id === submissionId) || {}
-    const studentId = submission.user_id || ''
-    const student = sections
-      .find(student => student.id === studentId) || {}
-    const studentName = student.name || ''
-    const studentNumber = student.sis_user_id || ''
-    const section = student.section || ''
+  const students = enrollments.filter(enrollment => enrollment.type === 'StudentEnrollment') // excludes StudentViewEnrollment but keeps StudentEnrollment, to keep studentView use enrollment.role
+  // const students = enrollments.filter(enrollment => enrollment.role === 'StudentEnrollment') use for testing
+  const nonStudents = enrollments.filter(enrollment => enrollment.role !== 'StudentEnrollment') // excludes both StudentViewEnrollment and StudentEnrollment
```
Comment (alisonmyers) [inline getRubric.js]: for testing use enrollment.role so that the test student is included

**Review by justin0022 — APPROVED** [review]

**Pull request merged**